lib/bundled_gems.rb

```diff
@@ -92,6 +94,13 @@ def self.find_gem(path)
   def self.warning?(name, specs: nil)
     # name can be a feature name or a file path with String or Pathname
     feature = File.path(name)
+
+    if feature.include?("/")
```

Comment:
Is it for full path by bootsnap?

Reply:
Yes, but even without bootsnap if you are requiring `some/thing`.

We could just always check `SINCE_FAST_PATH[File.basename(feature).sub(/(\.[^\.]+)\z/, "")]`, but this `include?` allow to skip the `File.basename` and `.sub`.

Reply:
I realized we could do:

```diff
diff --git a/lib/bundled_gems.rb b/lib/bundled_gems.rb
index f8f2ef5b27..245ff9b036 100644
--- a/lib/bundled_gems.rb
+++ b/lib/bundled_gems.rb
@@ -95,11 +95,7 @@ def self.warning?(name, specs: nil)
     # name can be a feature name or a file path with String or Pathname
     feature = File.path(name)
 
-    if feature.include?("/")
-      return unless SINCE_FAST_PATH[File.basename(feature).sub(/(\.[^\.]+)\z/, "")]
-    else
-      return unless SINCE_FAST_PATH[feature]
-    end
+    return unless SINCE_FAST_PATH[File.basename(feature, ".*")]
 
     # bootsnap expands `require "csv"` to `require "#{LIBDIR}/csv.rb"`,
     # and `require "syslog"` to `require "#{ARCHDIR}/syslog.so"`.
```

Which might be simpler.

Reply:
Is `File.basename(feature, ".*")` correct?
What if `feature` is e.g. `'rexml/node'` (or more realistic e.g. `digest/md5` but that's not extracted/warned)? Then `File.basename(feature, ".*")` is `node` and it will miss the warning, no?
I think we'd need the "root name" considering rubylibdir/rubyarchdir, i.e. similar to the logic below.

Reply:
It's an heuristic based on a few assumptions, the main one being that you'll require the entry point at some point. 

e.g. you can indeed `require "logger/formatter"` and that won't trigger a warnings, but the idea is that you'll very likely end up requiring `"logger"`.

In the `digest/md5` case it end up requiring `"digest"`:

- https://github.com/ruby/digest/blob/d152f5a043ea49ea7968bfebf7cd6b9318b4d557/ext/digest/md5/md5init.c#L55
- https://github.com/ruby/digest/blob/d152f5a043ea49ea7968bfebf7cd6b9318b4d557/ext/digest/digest.h#L57

But yes, it's possible there are a few corner cases not covered, but as explained in the issue, I think we need to be careful not to do too much work to identify the gem being called, because the performance of `require` is important.

Reply:
But if you find another heuristic that's more robust and in the same performance ballpark I'll be happy to review it.

Reply:
Could you document these assumptions as a code comment?
I don't see any mention of them before these 2 replies.